### ISSUE TITLE
Bug: [no-unsafe-type-assertion]: RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by hab25** [typescript-eslint/typescript-eslint#11824]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

```
Oops! Something went wrong! :(

ESLint: 9.39.1

RangeError: Maximum call stack size exceeded
Occurred while linting /home/hab25/my/work/typescript-eslint-issue/index.mts:24
Rule: "@typescript-eslint/no-unsafe-type-assertion"

...

```

### Reproduction Repository Link

https://github.com/hab25/typescript-eslint-issue

### Repro Steps

```bash
pnpm install && pnpm exec tsc && pnpm exec eslint --fix index.mts --no-cache --color --debug --max-warnings=0 --stats -c eslint.config.mjs
```

### Versions

```
$ node --version
v22.19.0
$ pnpm --version
10.22.0
$ pnpm list
[...]
devDependencies:
eslint 9.39.1
remeda 2.32.0
type-fest 5.3.0
typescript 5.9.3
typescript-eslint 8.48.1
```

**Comment by kirkwaiblinger** [comment]
Confirmed I can reproduce this... Not sure what the cause is but will probably need some investigation.

**Comment by mdm317** [comment]
[repro link](https://typescript-eslint.io/play/#ts=5.9.2&showAST=types&fileType=.tsx&code=C4TwDgpgBAggTAHgCpQgD2BAdgEwM5R7ABOAllgOYB8UAvFCuprgQAYAkA3uQGYTFQAYgF8uvflABKw1gCgoUAPxQOnEQCMu8BJKoz5UAFxQARCYDcs2aEhQAQokYZs%2BQiXLU6DVM5YqxWHwCIgFBUvoKyqoaXA46enIKxmaWsjgQAMYANgCGxNA8AK5YGcCkAPZYUDyOPsyuRGSUVAAUaMZIAJTGcUhUluhg5cTAUBmVRNXl5XBeyHUuBI0erep5HZ10NC01LWvEmzkE2n2dQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolpFStPsmSJo%2BShyboo06B2iRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

I believe this issue is caused by isTypeAssignableTo, and I’ve reported it as a bug to the TypeScript team.

**Comment by dbarabashh** [comment]
We could add a try/catch workaround to prevent the crash, but I'm wondering what is the approach here when typescript team eventually fixes this?

**Comment by kirkwaiblinger** [comment]
@dbarabashh If/when TS fixes https://github.com/microsoft/TypeScript/issues/62933 then the issue will just disappear whenever you upgrade your TS version.

I think we can use a try/catch for now. We've done something similar before with https://github.com/typescript-eslint/typescript-eslint/issues/10031

**Comment by dbarabashh** [comment]
@kirkwaiblinger okay, so the only follow-up would be eventually removing the try/catch as a cleanup once the supported typescript version includes the fix.

If that's okay, I can implement this.

**Comment by kirkwaiblinger** [comment]
@dbarabashh Yep, go for it!
